plenum/server/consensus/checkpoint_service.py

```diff
@@ -1,9 +1,351 @@
+import math
+from _sha256 import sha256
+from typing import Tuple
+
+from sortedcontainers import SortedListWithKey
+
+from common.exceptions import LogicError
+from common.serializers.serialization import serialize_msg_for_signing
+from plenum.common.config_util import getConfig
 from plenum.common.event_bus import InternalBus, ExternalBus
+from plenum.common.messages.internal_messages import StartMasterCatchup, StartBackupCatchup, Cleanup
+from plenum.common.messages.node_messages import Checkpoint, Ordered, CheckpointState
+from plenum.common.metrics_collector import MetricsName
+from plenum.common.stashing_router import DISCARD, PROCESS, StashingRouter
+from plenum.common.util import updateNamedTuple, SortedDict, firstKey
 from plenum.server.consensus.consensus_shared_data import ConsensusSharedData
+from plenum.server.consensus.msg_validator import CheckpointMsgValidator
+from plenum.server.replica import measure_replica_time, TPCStat, ConsensusDataHelper
+from stp_core.common.log import getlogger
 
 
 class CheckpointService:
-    def __init__(self, data: ConsensusSharedData, bus: InternalBus, network: ExternalBus):
+    STASHED_CHECKPOINTS_BEFORE_CATCHUP = 1
+
+    def __init__(self, data: ConsensusSharedData, bus: InternalBus, network: ExternalBus,
+                 stasher: StashingRouter, is_master=True):
         self._data = data
         self._bus = bus
         self._network = network
+        self._checkpoint_state = SortedDict(lambda k: k[1])
+        self._stasher = stasher
+        self._is_master = is_master
+        self._validator = CheckpointMsgValidator(self._data, self._checkpoint_state)
+        self._consensus_data_helper = ConsensusDataHelper(self._data)
```

Comment:
I think we need to get rid of using `ConsensusDataHelper` here and move checkpoint-related logic from there into CheckpointService